### PR TITLE
fix: don't raise item-click when focusable header is clicked after item keys are changed.

### PR DESCRIPTION
**Pull request by hdamr**
## Description
This PR fixes the issue when a focusable element is used in header (prepended), and clicking on this element causes `item-click` event to be raised (with a pre condition that grid item's `key` values are changed by something like filtering and clearing the filter after a grid row is clicked). It fixes it by using `getEventContext` and do more checking in order not to fire click event.
It also removed usage of `$connector.activeItem` which was introduced as a workaround [before](https://github.com/vaadin/vaadin-grid-flow/commit/74b944acb2da62bc7fffe106a7d8e93f8173ba4f).

Fixes [#2247](https://github.com/vaadin/flow-components/issues/2247)

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs-beta/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js [hunk 1]
```diff
@@ -1019,18 +1015,21 @@ import { ItemCache } from '@vaadin/grid/src/vaadin-grid-data-provider-mixin.js';
...
-            event.internalColumnId = eventContext.column._flowId;
-          }
-          grid.dispatchEvent(new CustomEvent(eventName,
-            { detail: event }));
+    function _fireClickEvent(event, eventName) {
+      const target = event.target;
+      const eventContext = grid.getEventContext(event);
+      const section = eventContext.section;
+
+      if (eventContext.item && !isFocusable(target) && section !== 'details') {
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1023]: Didn't test this yet, but wouldn't `!isFocusable(target)` mean that you'd no longer get an ItemClick event if you click a button on a row? Sounds like a breaking change if that's the case.

Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1023]: Wouldn't `section !== details` mean that you'd no longer get an ItemClick event if you click the row details? Sounds like a breaking change if that's the case.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js [hunk 2]
```diff
@@ -1019,18 +1015,21 @@ import { ItemCache } from '@vaadin/grid/src/vaadin-grid-data-provider-mixin.js';
...
+        event.itemKey = eventContext.item.key;
+        // if you have a details-renderer, getEventContext().column is undefined
+        if (eventContext.column) {
+          event.internalColumnId = eventContext.column._flowId;
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1027]: This check would be unnecessary if you couldn't enter this block if `section === 'details'`?

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js (hunk 1 above)
Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1023]: The previous workaround relied on the `cell-activate` event to determine which item to send an item click for. The `cell-activate` event it not dispatched when clicking details: https://github.com/vaadin/web-components/blob/master/packages/grid/src/vaadin-grid-active-item-mixin.js#L60-L62

There is also a Flow IT that covers that item-click is not triggered when clicking details.

Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1023]: The previous workaround relied on the `cell-activate` event to determine which item to send an item click for. The `cell-activate` event is not dispatched when clicking on a focusable element: https://github.com/vaadin/web-components/blob/master/packages/grid/src/vaadin-grid-active-item-mixin.js#L67

There is also a Flow IT that covers that item-click is not triggered when clicking a focusable element.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js [hunk 3]
```diff
@@ -1019,18 +1015,21 @@ import { ItemCache } from '@vaadin/grid/src/vaadin-grid-data-provider-mixin.js';
...
-        if (grid.$connector.activeItem) {
-          event.itemKey = grid.$connector.activeItem.key;
-          const eventContext = grid.getEventContext(event);
-          // if you have a details-renderer, getEventContext().column is undefined
-          if (eventContext.column) {
-            event.internalColumnId = eventContext.column._flowId;
-          }
-          grid.dispatchEvent(new CustomEvent(eventName,
-            { detail: event }));
+    function _fireClickEvent(event, eventName) {
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js:1018]: Please fix the indentation of the function to match the surrounding code.

**Review by sissbruecker — APPROVED** [review]
LGTM, let's wait for Tomi's approval

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js (hunk 1 above)
Comment (tomivirkki) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js]: Ah, got it

Comment (tomivirkki) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js]: Ah, got it

**Review by tomivirkki — APPROVED** [review]
LGTM also

**Pull request merged**